### PR TITLE
Fixing#427

### LINKED ISSUE #427: [REQUEST] Implement IfcPlanarBox
**Is your feature request related to a problem? Please describe.**
`IfcPlanarBox` is not supported. 

**Describe the solution you'd like**
Implement support in FaceConverter and call it appropriately from RepresentationConverter.

### PR DESCRIPTION
**Pull request by SamuilsRulovs**
Fixes #427

### CODE REVIEW

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 1]
```diff
@@ -1486,6 +1486,60 @@ namespace OpenInfraPlatform {
...
+					carve::math::Matrix inverseConicPositionMatrix = GeomUtils::computeInverse(conicPositionMatrix);
+
+
+					loops.push_back(inverseConicPositionMatrix * lowerLeftCorner);
+					loops.push_back(inverseConicPositionMatrix * lowerRightCorner);
+					loops.push_back(inverseConicPositionMatrix * upperRightCorner);
+					loops.push_back(inverseConicPositionMatrix * upperLeftCorner);
+
+					segmentStartPoints.push_back(inverseConicPositionMatrix * lowerLeftCorner);
+				}
```
Comment (SamuilsRulovs) [inline Core/src/IfcGeometryConverter/FaceConverter.h:1585]: I am not sure, If I have done it correctly, maybe I should do it like I did in IndexedPolygonalFace (create a triangulated list and push it to itemData)

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 2]
```diff
@@ -1486,6 +1486,60 @@ namespace OpenInfraPlatform {
...
+					std::vector<carve::geom::vector<3>>& loops,
+					std::vector<carve::geom::vector<3>>& segmentStartPoints)
+					const noexcept(false) 
+				{
+					double 	sizeInX = planarExtent->SizeInX;
+					double 	sizeInY = planarExtent->SizeInY;
+
+					//An extent is a rectangle specified by providing the coordinate of the lower left corner and the coordinate of the upper right corner in map units.
+					carve::geom::vector<3> lowerLeftCorner = carve::geom::VECTOR(-sizeInX, -sizeInY, 0.);
+					carve::geom::vector<3> upperRightCorner = carve::geom::VECTOR(sizeInX, sizeInY, 0.);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h]: This is wrong. Reading [here](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/FINAL/HTML/schema/ifcpresentationdefinitionresource/lexical/ifcplanarbox.htm): 

> The origin of this local coordinate system serves as the lower left corner of the rectangular box.

So `lowerLeftCorner` should be `(0,0)`.

File: Core/src/IfcGeometryConverter/FaceConverter.h [hunk 3]
```diff
@@ -1486,6 +1486,60 @@ namespace OpenInfraPlatform {
...
+					//I am not sure, if ItemData will take only 2 points of the rectangle, so I will add other 2 points here. 
+					carve::geom::vector<3> lowerRightCorner = carve::geom::VECTOR(sizeInX, -sizeInY, 0.);
+					carve::geom::vector<3> upperLeftCorner = carve::geom::VECTOR(-sizeInX, sizeInY, 0.);
+
+					//IfcPlanarBox
+					EXPRESSReference<typename IfcEntityTypesT::IfcPlanarBox> planarBox =
+						planarExtent.as<typename IfcEntityTypesT::IfcPlanarBox>();
+
+					carve::math::Matrix conicPositionMatrix = placementConverter->convertIfcAxis2Placement(planarBox->Placement);
+					carve::math::Matrix inverseConicPositionMatrix = GeomUtils::computeInverse(conicPositionMatrix);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/FaceConverter.h]: 1. No need for an inverse here.
1. Why call it `..conic..`?

**Review by pjanck — CHANGES_REQUESTED** [review]
> I am not sure, If I have done it correctly

Your superstition is correct. We need triangles here as well. Have a look at how `IfcPlane` is handled.

Additionally: this function only handles `IfcPlanarBox` (we don't need to handle `IfcPlanarExtent`, so the functionality suffices). Rename appropriately.

File: Core/src/IfcGeometryConverter/FaceConverter.h (hunk 2 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/FaceConverter.h]: Resolved: b4bb324

File: Core/src/IfcGeometryConverter/FaceConverter.h (hunk 3 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/FaceConverter.h]: Resolved: b4bb324

**Comment by SamuilsRulovs** [comment]
I guess this function is working correctly now

File: Core/src/IfcGeometryConverter/FaceConverter.h (hunk 1 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/FaceConverter.h:1585]: Resolved: b4bb324

**Review by pjanck — APPROVED** [review]
lftm

**Review by jschlenger — COMMENTED** [review]
Good work!
Can you add short doxy comments for the two functions you created and change the list of entities that we support so that ifcPlanarExtent and ifcPlanarBox are marked as supported?
Once you've done that I'll merge the pull request.

**Comment by SamuilsRulovs** [comment]
Ok, I completely forgot about that. I will do it ASAP

**Review by pjanck — APPROVED** [review]
👍

**Review by jschlenger — APPROVED** [review]

**Pull request merged**